breacharbiter.go

```diff
@@ -594,17 +596,287 @@ func (b *breachArbiter) createJusticeTx(r *retributionInfo) (*wire.MsgTx, error)
 	// witnesses for both commitment outputs, and all the pending HTLCs at
 	// this state in the channel's history.
 	// TODO(roasbeef): handle the 2-layer HTLCs
-	localWitness, err := r.selfOutput.witnessFunc(justiceTx, hashCache, 0)
+	localWitnessFunc := r.selfOutput.witnessType.GenWitnessFunc(
+		&b.wallet.Signer, r.selfOutput.signDescriptor)
+	localWitness, err := localWitnessFunc(justiceTx, hashCache, 0)
 	if err != nil {
 		return nil, err
 	}
 	justiceTx.TxIn[0].Witness = localWitness
 
-	remoteWitness, err := r.revokedOutput.witnessFunc(justiceTx, hashCache, 1)
+	remoteWitnessFunc := r.revokedOutput.witnessType.GenWitnessFunc(
+		&b.wallet.Signer, r.revokedOutput.signDescriptor)
+	remoteWitness, err := remoteWitnessFunc(justiceTx, hashCache, 1)
 	if err != nil {
 		return nil, err
 	}
 	justiceTx.TxIn[1].Witness = remoteWitness
 
 	return justiceTx, nil
 }
+
+// breachedOutput contains all the information needed to sweep a breached
+// output. A breached output is an output that we are now entitled to due to a
+// revoked commitment transaction being broadcast.
+type breachedOutput struct {
+	amt      btcutil.Amount
+	outpoint wire.OutPoint
+
+	signDescriptor *lnwallet.SignDescriptor
+	witnessType    lnwallet.WitnessType
+
+	twoStageClaim bool
+}
+
+// retribution encapsulates all the data needed to sweep all the contested
+// funds within a channel whose contract has been breached by the prior
+// counterparty. This struct is used to create the justice transaction which
+// spends all outputs of the commitment transaction into an output controlled
+// by the wallet.
+type retribution struct {
+	commitHash chainhash.Hash
+	chanPoint  wire.OutPoint
+
+	selfOutput    *breachedOutput
+	revokedOutput *breachedOutput
+	htlcOutputs   []*breachedOutput
+}
+
+// retributionStore handles persistence of retribution states to disk and is
+// backed by a boltdb bucket. The primary responsibility of the retribution
+// store is to ensure that we can recover from a restart in the middle of a
+// breached contract retribution.
+type retributionStore struct {
+	db *channeldb.DB
+}
+
+// newRetributionStore creates a new instance of a retributionStore.
+func newRetributionStore(db *channeldb.DB) *retributionStore {
+	return &retributionStore{
+		db: db,
+	}
+}
+
+// Add adds a retribution state to the retributionStore, which is then persisted
+// to disk.
+func (rs *retributionStore) Add(ret *retribution) error {
+	return rs.db.Update(func(tx *bolt.Tx) error {
+		// If this is our first contract breach, the retributionBucket won't
+		// exist, in which case, we just create a new bucket.
+		retBucket, err := tx.CreateBucketIfNotExists(retributionBucket)
+		if err != nil {
+			return err
+		}
+
+		var outBuf bytes.Buffer
+		if err := lnwire.WriteOutPoint(&outBuf, &ret.chanPoint); err != nil {
```

Comment:
By introducing the `WriteOutPoint` function within this PR, you've broken a few assumptions else where in the code base. Namely that the wire format differs from the on-disk format when storing various data. We strive to make them distinct so that modifying how we encode things on the wire doesn't also require us to modify our on-disk state. 

This is also the reason for one of the integration tests continually failing. Can you remove the introduction and usage of the `lnwire.WriteOutPoint`? Thanks!